Sim Control Flow Feature Addition

nruof opened the pull request:
Class that allows a user to define a list of function criteria to be evaluated over a set of diagnostic steps in a simulation.

Example:

steps_per_loop = 10
        while (sim_control.check_criteria()):
            sim.step(steps_per_loop)

nruof committed:
SimControl class to with terminate option on user defined criteria

nruof committed:
Moved sim_control.py out of diag_store, added a termination for max_loops

nruof committed:
Add max_loops termination

nruof committed:
Merge branch 'memaster' of https://github.com/ModernElectron/WarpX into nruof-sim-control

nruof committed:
Removed white space

KZhu-ME commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+        self.max_loops = max_loops
+        self.check_increment = 0
+
+    def add_checker(self, criteria=None, crit_args=None):
```
Not a big deal, but there doesn't seem to be a use case for where `criteria = None`, so `criteria` shouldn't have a default value of `None`.

KZhu-ME commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+        if crit_args:
+            self.crit_args_list.append(crit_args)
+        else:
+            self.crit_args_list.append({})
```
Here you don't need the if else statement/branching logic if you set the default value of `crit_args` to `crit_args={}`

peterscherpelz replied:
You actually don't want to do this as a general rule.
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html or a more verbose explanation 
https://towardsdatascience.com/python-pitfall-mutable-default-arguments-9385e8265422

KZhu-ME commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+    def check_criteria(self):
+        """
+        Evaluates each criteria in the crit_list
+        returns True is all criteria are True otherwise returns False
```
```suggestion
        returns True if all criteria are True otherwise returns False
```
A small typo

KZhu-ME reviewed (commented):
Looks good, just a have a few minor suggestions.

roelof-groenewald commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+    set of user defined functions and criteria
+
+    """
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
```
This class should definitely allow for the scenario where `max_loops` is the only argument specified. So I'd suggest giving default values of `None` for `criteria` and `criteria_args` and not setting a default value for max_loops. Then add check in the __init__ function that max_loops >= 1.

peterscherpelz replied:
I would not separate criteria and criteria_args into separate lists. (I don't like the format of having to match up correct criteria with their args.)
A nice format might be a list of criteria, and each criteria can either be a function or a tuple of `(function, kwargs)`.
```
if callable(criteria[i]):
   self.crit_list.append(criteria[i])
   self.crit_args_list.append({})
else:
   self.crit_list.append(criteria[i][0])
   self.crit_args_list.append(criteria[i][1])
```

peterscherpelz commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+
+# Get module-level logger
+logger = logging.getLogger(__name__)
+
```
```suggestion


```

roelof-groenewald commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+            sim.step(steps_per_loop)
+
+        """
+        if self.check_increment > 0:
+            keep_going = False
+            for i, func in enumerate(self.crit_list):
+                keep_going = self.crit_list[i](**self.crit_args_list[i])
+                if not keep_going:
+                    print(("Termination due to criteria {}").format(func.__name__))
+                    return keep_going
+            self.check_increment += 1
+            if self.max_loops:
+                if self.check_increment >= self.max_loops:
+                    keep_going = False
+            return keep_going
+        else:
+            self.check_increment += 1
+            return True
```
With the forcing of a `max_loops` value from above I'd change this function to something like
```suggestion
        self.check_increment += 1
        if self.check_increment >= self.max_loops:
            print('Max loops reached.')
            return False
            
        for i, func in enumerate(self.crit_list):
            if not self.crit_list[i](**self.crit_args_list[i]):
                print(("Termination due to criteria {}").format(func.__name__))
                return False
```
But actually even better than using `max_loops` argument would be to use a `max_steps` argument and then calling the mewarpx function that returns the simulation step number when evaluating whether to terminate.

peterscherpelz commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+"""
+
+import logging
+import time
```
Remove unused import

peterscherpelz replied:
and numpy, psutil, _libwarpx, mwxrun - or at least comment out if they'll be later used as you expand this.

roelof-groenewald reviewed (changes requested):
I left some comments on suggested changes, please let me know if anything is unclear.

peterscherpelz commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
+        """
+        Generate and install functions to perform after a step #.
+        Arguments:
```
Newline before Arguments will make autodoc happier. Then indent criteria and criteria_args by 4 spaces relative to arguments.
Also add max_loops and kwargs to Arguments list
You should either say 'read check_criteria' documentation below, or move info about criteria up here - eg when do we want it to return True vs False.

peterscherpelz commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+
+        criteria: list of user defined functions that return True or False value
+
+        criteria_args: list of kwargs for each criteria
```
dict, not list

peterscherpelz commented on mewarpx/mewarpx/sim_control.py:
```diff
@@ -0,0 +1,71 @@
+            keep_going = False
+            for i, func in enumerate(self.crit_list):
+                keep_going = self.crit_list[i](**self.crit_args_list[i])
+                if not keep_going:
```
I would not short-circuit these function evaluations if one returns false. Sometimes we'll want to install functions that both handle diagnostics and tell something to terminate. It'll probably be most straightforward if those all still run on the last time through things.

peterscherpelz replied:
You can keep the same print info by starting with an empty string, then add to it for each criteria that returns False.

peterscherpelz reviewed (changes requested):
Lots of comments, mostly for writing clean code in the long-term. Thanks!

nruof committed:
Kevin and Roelof changes

nruof committed:
Change max_loops to max_steps

nruof committed:
Peter's changes and testing, debug

KZhu-ME reviewed (approved):
Aside from the changes Peter suggested, everything else looks good to me!

EtherealMC-Bit reviewed (approved):
Cool stuff

peterscherpelz committed:
Tweak documentation and formatting

Note especially empty line before `Arguments:`, and indentation of these
blocks. Also, the `::` after "Example Sim run" makes the indented text
display as a code block.

peterscherpelz committed:
Remove unused **kwargs

peterscherpelz committed:
Eliminate duplicate functionality, doc tweaks

- initialize_criteria calls add_checker to avoid repeating it
- max_steps is still required, but now implemented as a criteria similar
  to the rest
- Note this also means max_steps doesn't short-circuit all other
  criteria being run
- Add types to Arguments docs; break long line

nruof commented:
I tested out the changes.  eval_max_steps terminates the simulation properly a custom criteria function also terminates properly.  Changes are good.

roelof-groenewald reviewed (approved):
Looks good to me.
